### PR TITLE
Rework handling of /hub/user/:name

### PR DESCRIPTION
**Pull request by minrk**
... to do less.

This mostly serves to simplify things and avoid implicit spawns, especially when it comes to API requests to a not-running server

- [x] entirely avoid implicit spawn on /hub/user/:name. Visiting this URL no longer has any side effects. It:
- [x] API requests to not-running servers,
  as detected when the verb is not GET,
  or `Accept` specifies application/json,
  raise 503 with an informative JSON error message.
- [x] new /hub/spawn-pending/:user/:name URL that serves the progress page. Visits to /hub/user/:name for a server pending launch will be redirected here.
- [x] browser request to /user/:name/:server if it's not running renders
  a page (status 503) with a single button saying that the server is not running,
  click the link to spawn. This puts the old behavior a single click away
  for users.
- [x] Spawn links in UI should all point to /hub/spawn/:user/:server instead of /user/:user. This fixes several cases
  where the options form would not be rendered for named servers
- [x] Detect server name in links to not running /user/:name/:server/...
- [x] includes and finishes #2341
- [x] update tests
- [x] tests for new spawn-pending page

Since this updates all kinds of redirects and events during spawn, several page tests are going to need updating. The API is unaffected.

### CODE REVIEW

**Comment by minrk** [comment]
@yuvipanda this essentially commits even more strongly the behavior that prompted #2293, I think putting increased pressure on the solution being JupyterLab drawing a dialog with a link to restart the server. I opened https://github.com/jupyterhub/jupyterlab-hub/issues/83 to track that.

**Comment by yuvipanda** [comment]
Thanks, @minrk! I agree this puts more pressure on client side tools. We should be able to hopefully do this without having to have users refresh the page - just a kernel restart instead.

We should also do this for classic notebook. Would that have to be an nbextension? I wonder if we can automatically inject that from jupyterhub-singleuser

**Comment by minrk** [comment]
OK, after chasing down all of the test errors (almost all had to do with tests assuming when requests to `/hub/spawn` returned a server was ready, which is now rarely the case as /hub/spawn returns immediately, sending users to /spawn-pending/ to wait)

This is the last thing I wanted for 1.0, so I think we can start the release process if folks want to give this a review.

File: jupyterhub/handlers/base.py [hunk 1]
```diff
@@ -1127,27 +1140,87 @@ def get(self):
         self.redirect(url_path_join(self.hub.base_url, path), permanent=False)
 
 
-class UserSpawnHandler(BaseHandler):
-    """Redirect requests to /user/user_name/* handled by the Hub.
+class UserUrlHandler(BaseHandler):
+    """Handle requests to /user/user_name/* routed to the Hub.
+
+    As of 1.0, this does *not* trigger a spawn any more. Instead, it:
```
Comment (willingc) [inline jupyterhub/handlers/base.py]:
```suggestion
    **Changed Behavior as of 1.0** This handler no longer triggers a spawn. Instead, it checks if:
```

File: jupyterhub/handlers/base.py [hunk 2]
```diff
@@ -1166,209 +1239,162 @@ def _fail_api_request(self, user):
             )
             admin_spawn = True
             should_spawn = True
+            redirect_to_self = False
         else:
             user = current_user
             admin_spawn = False
-            # For non-admins, we should spawn if the user matches
+            # For non-admins, we should spawn if the user requested is the current usr
```
Comment (willingc) [inline jupyterhub/handlers/base.py]:
```suggestion
            # For non-admins, spawn if the user requested is the current user
```

File: share/jupyterhub/templates/home.html [hunk 3]
```diff
@@ -72,6 +72,7 @@ <h2>
         <td>
           <a role="button" class="stop-server btn btn-xs btn-danger{% if not spawner.active %} hidden{% endif %}" id="stop-{{ spawner.name }}">stop</a>
           <a role="button" class="start-server btn btn-xs btn-primary {% if spawner.active %} hidden{% endif %}" id="start-{{ spawner.name }}"
+          href="{{ base_url }}spawn/{{ user.name }}/{{ spawner.name }}"
```
[the remark this answers is not in the document]
Comment (minrk) [reply share/jupyterhub/templates/home.html]: yes, we make sure that base_url is guaranteed to start and end with slash

File: jupyterhub/handlers/pages.py [hunk 4]
```diff
@@ -56,9 +61,9 @@ class HomeHandler(BaseHandler):
         # to establish that this is an explicit spawn request rather
         # than an implicit one, which can be caused by any link to `/user/:name(/:server_name)`
         url = (
```
[the remark this answers is not in the document]
Comment (minrk) [reply jupyterhub/handlers/pages.py]: Yup, much better!

**Comment by minrk** [comment]
Thanks for the review, @willingc! I think it's all addressed now.

**Comment by willingc** [comment]
Thanks @minrk. Let's see how this works out in the wild now :wink: Awesome job. 

Will continue poking at the tests. It looks like that local failures/errors are due to mocking.

**Pull request merged**